File: MovieInfo/Assign8IOS/Assign8IOS/VideoPlayer.swift
/* Copyright 2016 Viplav Kadam
* Licensed under the Apache License, Version 2.0 (the "License");
* you may not use this file except in compliance with the License.
* You may obtain a copy of the License at
*
* http://www.apache.org/licenses/LICENSE-2.0
*
* Unless required by applicable law or agreed to in writing, software
* distributed under the License is distributed on an "AS IS" BASIS,
* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
* See the License for the specific language governing permissions and
* limitations under the License.
*
* Purpose: An iOS app for video playback by download to local file, and also add movie info from IMDB API.
*
* This can be used by
* @Professor Tim Lindquist
* @University Arizona State University
* @author Viplav Kadam mailto: vckadam@asu.edu
* @version 26 April 2016
*/

import UIKit
import AVKit
import AVFoundation

class VideoPlayer: AVPlayerViewController, NSURLSessionDelegate {
    
    var filename:String?
    
    var streamer_host:String?
    
    var streamer_port:String?
    
    override func viewDidLoad() {
        
        super.viewDidLoad()
        
        let value = UIInterfaceOrientation.LandscapeRight.rawValue
        
        UIDevice.currentDevice().setValue(value, forKey:"orientation")
        
        self.navigationController?.navigationBarHidden = true
        
        
        if let path = NSBundle.mainBundle().pathForResource("Server", ofType: "plist"){
            
            if let dict = NSDictionary(contentsOfFile: path) as? [String:AnyObject] {
                streamer_host = dict["host"] as? String
                streamer_port = dict["streamer_port"] as? String
            }
            let urlString:String = "http://\(streamer_host!):\(streamer_port!)/"+self.filename!;
            
            downloadVideo(urlString)
        }
        
    }
    
    
    override func viewWillDisappear(animated: Bool){
        
        if let status:AVPlayerStatus = self.player?.status {
            
            NSLog("viewWillDisappear \(((status==AVPlayerStatus.ReadyToPlay) ? "Ready":"unknown")))")
            
        }else{
            
            NSLog("viewWillDisappear player not initialized")
            
        }
        
        if self.player != nil {
            
            self.player?.pause()
            
        }
        self.dismissViewControllerAnimated(true, completion: nil)
        
        let value = UIInterfaceOrientation.Portrait.rawValue
        
        UIDevice.currentDevice().setValue(value, forKey:"orientation")
        
        self.navigationController?.navigationBarHidden = false
        
    }
    
    
    override func didReceiveMemoryWarning() {
        
        super.didReceiveMemoryWarning()
        
    }
    
    
    func downloadVideo(urlString: String){
        
        let bgConf = NSURLSessionConfiguration.backgroundSessionConfigurationWithIdentifier("bgSession")
        
        let backSess = NSURLSession(configuration: bgConf, delegate: self, delegateQueue:NSOperationQueue.mainQueue())
        
        let aUrl = NSURL(string: urlString)!
        
        let downloadBG = backSess.downloadTaskWithURL(aUrl)
        
        downloadBG.resume()
    }
    
    
    func playMovieAtURL(fileURL: NSURL){
        
        if (self.player != nil && self.player!.status == AVPlayerStatus.ReadyToPlay) {
            
            let playerItem = AVPlayerItem(URL: fileURL)
            
            self.player?.replaceCurrentItemWithPlayerItem(playerItem)
            
        }else{
            
            self.player = AVPlayer(URL: fileURL)
            
        }
        self.videoGravity = AVLayerVideoGravityResizeAspectFill
        
        self.player!.play()
    }
    
    
    func URLSession(session: NSURLSession,
        downloadTask: NSURLSessionDownloadTask,
        didFinishDownloadingToURL location: NSURL){
            
            let path = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true)
            
            let documentDirectoryPath:String = path[0]
            
            let fileMgr = NSFileManager()
            
            let destinationURLForFile = NSURL(fileURLWithPath: documentDirectoryPath.stringByAppendingString("/MinionsBanana.mp4"))
            
            if fileMgr.fileExistsAtPath(destinationURLForFile.path!) {
                
                NSLog("destination file url: \(destinationURLForFile.path!) exists. Deleting")
                
                do {
                    
                    try fileMgr.removeItemAtURL(destinationURLForFile)
                    
                }catch{
                    
                    NSLog("error removing file at: \(destinationURLForFile)")
                    
                }
            }
            do {
                
                try fileMgr.moveItemAtURL(location, toURL: destinationURLForFile)
                
                NSLog("download and save completed to: \(destinationURLForFile.path!)")
                
                session.invalidateAndCancel()
                
                playMovieAtURL(destinationURLForFile)
                
            }catch{
                
                NSLog("An error occurred while moving file to destination url")
                
            }
    }
    
    func URLSession(session: NSURLSession,
        downloadTask: NSURLSessionDownloadTask,
        didWriteData bytesWritten: Int64,
        totalBytesWritten: Int64,
        totalBytesExpectedToWrite: Int64){
            
            NSLog("did write portion of file: \(Float(totalBytesWritten)/Float(totalBytesExpectedToWrite))")
            
    }
    
}

File: MovieInfo/MyAssing3Android/app/src/main/java/edu/asu/msse/vckadam/myassing3android/VideoPlayer.java
package edu.asu.msse.vckadam.myassing3android;

/*
* Copyright 2016 Viplav Kadam
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Purpose : To create which support storing movie description from OMDB as well as local server
 * and also plays video file
 *
 * This can be used by
 * @Professor Tim Lindquist
 * @University Arizona State University
 * @author Viplav Kadam mailto: vckadam@asu.edu
 * @version 26 April 2016
 */
import android.content.Context;
import android.content.Intent;
import android.content.res.Configuration;
import android.media.MediaMetadataRetriever;
import android.media.MediaPlayer;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.widget.MediaController;
import android.widget.VideoView;

public class VideoPlayer extends AppCompatActivity implements MediaPlayer.OnPreparedListener {


    private VideoView mVideoView;

    private MediaController mController;

    private MediaMetadataRetriever mMetadataRetriever;

    String filename;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_video_player);

        Intent i = getIntent();

        filename = i.getStringExtra("filename");

        mVideoView = (VideoView) findViewById(R.id.avideoview);

        mVideoView.setVideoPath(getString(R.string.urlprefix) + filename);


        MediaController mediaController = new MediaController((Context)this);

        mediaController.setAnchorView(mVideoView);

        mVideoView.setMediaController(mediaController);

        mVideoView.setOnPreparedListener(this);
    }

    public void onPrepared(MediaPlayer mp){
        android.util.Log.d(this.getClass().getSimpleName(), "onPrepared called. Video Duration: "
                + mVideoView.getDuration());
        mVideoView.start();
    }

    @Override
    public void onConfigurationChanged(Configuration newConfig) {
        android.util.Log.d(this.getClass().getSimpleName(), "onConfigurationChanged");
        super.onConfigurationChanged(newConfig);
    }
}
